src/transformers/models/llama/modeling_llama.py

```diff
@@ -815,9 +815,13 @@ def _setup_cache(self, cache_cls, max_batch_size, max_cache_len: Optional[int] =
             self.register_buffer("causal_mask", torch.triu(causal_mask, diagonal=1), persistent=False)
 
         for layer in self.model.layers:
-            weights = layer.self_attn.o_proj.weight
+            device = layer.input_layernorm.weight.device
+            if hasattr(self.config, "_pre_quantization_dtype"):
+                dtype = self.config._pre_quantization_dtype
+            else:
+                dtype = layer.self_attn.o_proj.weight.dtype
```

Comment:
```suggestion
            dtype = layer.self_attn.o_proj.weight.dtype
            if hasattr(self.config, "_pre_quantization_dtype"):
                dtype = self.config._pre_quantization_dtype
```
not super fan of having this in the modeling file, but that's alright!

Reply:
@ArthurZucker many quantized linear layers don't have a `weight` attribute at all. This will raise the same error as [described above](https://github.com/huggingface/transformers/pull/29079#issuecomment-1959093412).

Reply:
Ah yeah right feel free to ignore then 👍🏻